store/store_download.go

```diff
@@ -308,6 +309,62 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 	return s.cacher.Put(downloadInfo.Sha3_384, targetPath)
 }
 
+var errIconUnchanged = errors.New("existing icon unchanged")
+
+// DownloadIcon downloads the icon for the snap from the given download URL to
+// the given target path. Snap icons are small (<256kB) files served from an
+// ordinary unauthenticated file server, so this does not require store
+// authentication or user state, nor a progress bar. They are also not revision-
+// specific, and do not use a download cache.
+func DownloadIcon(ctx context.Context, name string, targetPath string, downloadURL string) error {
+	if err := os.MkdirAll(filepath.Dir(targetPath), 0o755); err != nil {
+		return err
+	}
+
+	// Read etag of existing file at targetPath, if it exists
+	var etag string
+	etagBuf := make([]byte, 256) // all etags should be smaller than 256B
+	if size, err := unix.Getxattr(targetPath, "user.etag", etagBuf); err == nil {
+		etag = string(etagBuf[:size])
+	}
+
+	// Download icon to a temporary file location, since there may be an active
+	// snap icon hard-linked to targetPath. If download fails, don't want to
+	// corrupt the existing icon file contents. Instead, commit the new file
+	// once the download succeeds, thus leaving any previous icon linked to the
+	// original unchanged icon contents, until a future task re-links to the
+	// new contents.
+	aw, err := osutil.NewAtomicFile(targetPath, 0o644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return fmt.Errorf("cannot create file for snap icon for snap %s: %v", name, err)
+	}
+	// on success, Cancel becomes a no-op
+	defer aw.Cancel()
+
+	logger.Debugf("Starting download of %q to %q.", downloadURL, targetPath)
+
+	etag, err = downloadIcon(ctx, name, etag, downloadURL, aw)
+	if errors.Is(err, errIconUnchanged) {
+		logger.Debugf("download of snap icon skipped for snap %s: icon unchanged", name)
+		return nil
+	} else if err != nil {
+		logger.Debugf("download of %q failed: %#v", downloadURL, err)
+		return err
+	}
+
+	if err = aw.Commit(); err != nil {
+		return fmt.Errorf("cannot commit snap icon file for snap %s: %v", name, err)
+	}
+
+	// Success, now try to store the etag
+	if etag != "" {
+		// Ignore any error. If it fails, we'll just redownload the whole icon
+		// next time. No problem.
+		unix.Setxattr(targetPath, "user.etag", []byte(etag), 0)
```

Comment:
```suggestion
		// Ignore any error in case the filesystem does not support xattrs.
		// If it fails, we'll just redownload the whole icon next time
		unix.Setxattr(targetPath, "user.snapstore-etag", []byte(etag), 0)
```